### ISSUE TITLE
cmd/compile: atomic intrinsics for runtime and sync/atomic

### ISSUE THREAD

**Issue by dvyukov** [golang/go#8739]
<pre>We need to unify interface of atomic operations between runtime package and sync/atomic
package. And then teach gc to replace functions in both packages with intrinsics where
possible.

This is followup to <a href="https://golang.org/issue/4947">https://golang.org/issue/4947</a> which was about
atomic intrinsics for runtime C code.</pre>

**Comment by rsc** [comment]
There's no evidence in profiles that this is important for Go 1.5.

**Comment by dsnet** [comment]
\cc @randall77, I vaguely remember you mentioning that we already intrinsify atomic.X functions. Is this done?

**Comment by randall77** [comment]
They are for 1.8.  We intrinsify both sync/atomic and runtime/internal/atomic for a bunch of architectures.
The APIs are not unified (e.g. LoadUint32 in sync/atomic is Load in runtime/internal/atomic).
We should probably leave this bug open for that unification.  It shouldn't be hard.

**Comment by laboger** [comment]
I'm planning to implement the atomics as intrinsics on ppc64x.  Should I refer to this issue in my CL or open a new one.

**Comment by randall77** [comment]
You can reference this issue.

**Comment by laboger** [comment]
Keith, I've had a change for this that worked fine a few months ago.  When I tried to rebase it, now it hangs during the build.  I've tracked it down to commit 8a9dc05 to allow inlining of functions with intrinsics in them.  If I remove the code from that commit, then my change to intrinsify atomics works again.

**Comment by randall77** [comment]
@laboger Can you point me at your intrinsics CL?

**Comment by laboger** [comment]
@randall77 I didn't create one yet since it didn't build.  But I will do that now.

**Comment by laboger** [comment]
CL #36832

**Comment by mark-rushakoff** [comment]
Was this issue supposed to be closed with CL 36832? The CL is specific to PPC64x but it looks like this issue is not specific to any architecture.

**Comment by randall77** [comment]
No, we should leave it open.  I think the remaining task is unifying the apis for sync/atomic and runtime/internal/atomic.
